rice-datanode/source/data_transfer_server.cc

```diff
@@ -485,18 +469,48 @@ bool TransferServer::replicate(uint64_t len, std::string ip,
   }
   xmits--;
   cv.notify_one();
+  return true;
+}
+
+bool TransferServer::replicate(uint64_t len, std::string ip,
+                               std::string xferport,
+                               ExtendedBlockProto blockToTarget) {
+  LOG(INFO)
+    << " replicating length "
+    << len
+    << " with ip "
+    << ip
+    << " and port "
+    << xferport;
+
+  LOG(INFO) << "blockToTarget is " << blockToTarget.blockid();
+
+  if (fs->hasBlock(blockToTarget.blockid())) {
+    LOG(INFO) << "Block already exists on this DN";
+    return true;
+  } else {
+    LOG(INFO) << "Block not found on this DN, replicating...";
+  }
+
+  std::string data(len, 0);
+  int read_len = 0;
+  if (!remote_read(len, ip, xferport, blockToTarget, data, read_len)) {
+    return false;
+  }
+  uint64_t block_id = blockToTarget.blockid();
 
   // TODO(anyone): send ClientReadStatusProto (delimited)
-  if (!fs->writeBlock(header->baseheader().block().blockid(), data)) {
+  if (!fs->writeBlock(block_id, data)) {
     LOG(ERROR)
       << "Failed to allocate block "
-      << header->baseheader().block().blockid();
+      << block_id;
     return false;
   } else {
-    dn->blockReceived(header->baseheader().block().blockid(), read_len);
+    dn->blockReceived(block_id, read_len);
   }
 
-  base_proto->release_block();
+  // TODO(ADAM): Make sure this line is unecessary
```

Comment:
is there anything preventing us from doing this now

Reply:
From everything I've looked at, I'm pretty sure it's unnecessary, but I want to keep it marked just in case. Updating comments to reflect this.